---
title: "Collapse barcodes to final per-RBD/mutant phenotype scores for the wildtype sarbecoviruses pool"
author: "Tyler Starr"
date: "9/27/2023"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---
This notebook reads in the per-barcode sera binding values and previously measured expression for sarbecovirus homologs pool. It synthesizes these two sets of results and calculates the final 'mean' phenotypes for each variant, and generates some coverage and QC analyses.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages],
                   lib=c(paste("/uufs/chpc.utah.edu/common/home/",Sys.getenv("USER"),"/RLibs/",Sys.getenv("R_VERSION"),sep="")),
                   repos=c("http://cran.us.r-project.org"))
}
#load packages
invisible(lapply(packages, library, character.only=T))

knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$final_variant_scores_dir)){
  dir.create(file.path(config$final_variant_scores_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Setup

Read in tables of per-barcode AUC

```{r input_data}
dt <- data.table(read.csv(config$mAb_EC50_file),stringsAsFactors=F)[library=="lib61_SARSr-wts",]

#check all targets are in the targets_ordered config list
unique(dt$target) %in% config$targets_ordered

#assign target as a factor in my desired overall plotting order
dt[,target := factor(dt$target,levels=config$targets_ordered)]

#read in previously measured expression measurements for these variants? Note, don't have this metric for the sarbs that have been pooled in since the original library pool back in 2021
dt_expr <- data.table(read.csv(config$SARSr_lib61_mut_bind_expr),stringsAsFactors=F)
dt_expr[target=="SARS-CoV-2",target:="SARS-CoV-2_WH1"]
```

## Calculate per-variant mean scores

Unfiltered, look at distribution of AUC scores

```{r unfiltered_AUCs, echo=T, fig.width=10, fig.height=50, fig.align="center", dpi=300,dev="png"}
p1 <- ggplot(dt[!is.na(EC50_S2V29_v37_2),],aes(x=target,y=log10(EC50_S2V29_v37_2)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29_v37_2 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p2 <- ggplot(dt[!is.na(EC50_S2V29),],aes(x=target,y=log10(EC50_S2V29)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p3 <- ggplot(dt[!is.na(EC50_S2V29_UCA),],aes(x=target,y=log10(EC50_S2V29_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p4 <- ggplot(dt[!is.na(EC50_S2K146_v860),],aes(x=target,y=log10(EC50_S2K146_v860)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146_v860 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p5 <- ggplot(dt[!is.na(EC50_S2K146),],aes(x=target,y=log10(EC50_S2K146)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p6 <- ggplot(dt[!is.na(EC50_S2K146_UCA),],aes(x=target,y=log10(EC50_S2K146_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p7 <- ggplot(dt[!is.na(EC50_S3L17),],aes(x=target,y=log10(EC50_S3L17)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S3L17 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p8 <- ggplot(dt[!is.na(EC50_S3L17_UCA),],aes(x=target,y=log10(EC50_S3L17_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S3L17_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p9 <- ggplot(dt[!is.na(EC50_S309),],aes(x=target,y=log10(EC50_S309)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S309 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p10 <- ggplot(dt[!is.na(EC50_S309_887),],aes(x=target,y=log10(EC50_S309_887)))+
   geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
   ggtitle("S309_887 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p11 <- ggplot(dt[!is.na(EC50_S2X259),],aes(x=target,y=log10(EC50_S2X259)))+
   geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
   ggtitle("S2X259 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p12 <- ggplot(dt[!is.na(EC50_SA55),],aes(x=target,y=log10(EC50_SA55)))+
   geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
   ggtitle("SA55 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p13 <- ggplot(dt[!is.na(EC50_Omi42),],aes(x=target,y=log10(EC50_Omi42)))+
   geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
   ggtitle("Omi42 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

grid.arrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13,ncol=1)
```

Let's add a variable that flags the top and bottom 5% of expression scores for each variant, and see how violin plots look when censoring these top and bottom 10% of expressed barcodes. Let's *only* do this for targets that are in the many-barcode pool, not the spike-in targets that had between 2-5 barcodes total (but with much higher per-barcode cell counts for the titraitons!)

```{r censor_5_AUCs, echo=T, fig.width=10, fig.height=50, fig.align="center", dpi=300,dev="png"}
dt[,`S2V29_v37_2_censor_lower`:=quantile(EC50_S2V29_v37_2,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2V29_v37_2_censor_upper`:=quantile(EC50_S2V29_v37_2,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2V29_censor_lower`:=quantile(EC50_S2V29,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2V29_censor_upper`:=quantile(EC50_S2V29,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2V29_UCA_censor_lower`:=quantile(EC50_S2V29_UCA,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2V29_UCA_censor_upper`:=quantile(EC50_S2V29_UCA,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2K146_v860_censor_lower`:=quantile(EC50_S2K146_v860,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2K146_v860_censor_upper`:=quantile(EC50_S2K146_v860,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2K146_censor_lower`:=quantile(EC50_S2K146,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2K146_censor_upper`:=quantile(EC50_S2K146,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2K146_UCA_censor_lower`:=quantile(EC50_S2K146_UCA,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2K146_UCA_censor_upper`:=quantile(EC50_S2K146_UCA,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S3L17_censor_lower`:=quantile(EC50_S3L17,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S3L17_censor_upper`:=quantile(EC50_S3L17,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S3L17_UCA_censor_lower`:=quantile(EC50_S3L17_UCA,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S3L17_UCA_censor_upper`:=quantile(EC50_S3L17_UCA,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S309_censor_lower`:=quantile(EC50_S309,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S309_censor_upper`:=quantile(EC50_S309,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S309_887_censor_lower`:=quantile(EC50_S309_887,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S309_887_censor_upper`:=quantile(EC50_S309_887,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`S2X259_censor_lower`:=quantile(EC50_S2X259,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`S2X259_censor_upper`:=quantile(EC50_S2X259,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`SA55_censor_lower`:=quantile(EC50_SA55,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`SA55_censor_upper`:=quantile(EC50_SA55,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`Omi42_censor_lower`:=quantile(EC50_Omi42,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`Omi42_censor_upper`:=quantile(EC50_Omi42,0.95,na.rm=T,type=7),by=c("library","target")]

p1 <- ggplot(dt[!is.na(EC50_S2V29_v37_2) & ((EC50_S2V29_v37_2 > S2V29_v37_2_censor_lower & EC50_S2V29_v37_2 < S2V29_v37_2_censor_upper) | target %in% config$targets_low_bc) ,],
             aes(x=target,y=log10(EC50_S2V29_v37_2)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29_v37_2 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p2 <- ggplot(dt[!is.na(EC50_S2V29) & ((EC50_S2V29 > S2V29_censor_lower & EC50_S2V29 < S2V29_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2V29)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p3 <- ggplot(dt[!is.na(EC50_S2V29_UCA) & ((EC50_S2V29_UCA > S2V29_UCA_censor_lower & EC50_S2V29_UCA < S2V29_UCA_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2V29_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2V29_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p4 <- ggplot(dt[!is.na(EC50_S2K146_v860) & ((EC50_S2K146_v860 > S2K146_v860_censor_lower & EC50_S2K146_v860 < S2K146_v860_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2K146_v860)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146_v860 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p5 <- ggplot(dt[!is.na(EC50_S2K146) & ((EC50_S2K146 > S2K146_censor_lower & EC50_S2K146 < S2K146_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2K146)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p6 <- ggplot(dt[!is.na(EC50_S2K146_UCA) & ((EC50_S2K146_UCA > S2K146_UCA_censor_lower & EC50_S2K146_UCA < S2K146_UCA_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2K146_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2K146_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p7 <- ggplot(dt[!is.na(EC50_S3L17) & ((EC50_S3L17 > S3L17_censor_lower & EC50_S3L17 < S3L17_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S3L17)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S3L17 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p8 <- ggplot(dt[!is.na(EC50_S3L17_UCA) & ((EC50_S3L17_UCA > S3L17_UCA_censor_lower & EC50_S3L17_UCA < S3L17_UCA_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S3L17_UCA)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S3L17_UCA mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p9 <- ggplot(dt[!is.na(EC50_S309) & ((EC50_S309 > S309_censor_lower & EC50_S309 < S309_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S309)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S309 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p10 <- ggplot(dt[!is.na(EC50_S309_887) & ((EC50_S309_887 > S309_887_censor_lower & EC50_S309_887 < S309_887_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S309_887)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S309_887 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p11 <- ggplot(dt[!is.na(EC50_S2X259) & ((EC50_S2X259 > S2X259_censor_lower & EC50_S2X259 < S2X259_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_S2X259)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("S2X259 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p12 <- ggplot(dt[!is.na(EC50_SA55) & ((EC50_SA55 > SA55_censor_lower & EC50_SA55 < SA55_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_SA55)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("SA55 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p13 <- ggplot(dt[!is.na(EC50_Omi42) & ((EC50_Omi42 > Omi42_censor_lower & EC50_Omi42 < Omi42_censor_upper) | target %in% config$targets_low_bc),],
             aes(x=target,y=log10(EC50_Omi42)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("Omi42 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))


grid.arrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13,ncol=1)
invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_vioplots_AUC-cens.pdf",sep="")))
```

Calculate the mean (geometric mean) per variant, the standard deviation, and the number of (post-filter) barcodes on which a variant score was determined

```{r calculate_mean_scores}
#apply the censors to NA out the phenotypes outside the range for the targets that aren't in the 2-5 higher-count barcode spike in list
dt[!(target %in% config$targets_low_bc) & (EC50_S2V29_v37_2 < `S2V29_v37_2_censor_lower` | EC50_S2V29_v37_2 > `S2V29_v37_2_censor_upper`), EC50_S2V29_v37_2:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2V29 < `S2V29_censor_lower` | EC50_S2V29 > `S2V29_censor_upper`), EC50_S2V29:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2V29_UCA < `S2V29_UCA_censor_lower` | EC50_S2V29_UCA > `S2V29_UCA_censor_upper`), EC50_S2V29_UCA:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2K146_v860 < `S2K146_v860_censor_lower` | EC50_S2K146_v860 > `S2K146_v860_censor_upper`), EC50_S2K146_v860:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2K146 < `S2K146_censor_lower` | EC50_S2K146 > `S2K146_censor_upper`), EC50_S2K146:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2K146_UCA < `S2K146_UCA_censor_lower` | EC50_S2K146_UCA > `S2K146_UCA_censor_upper`), EC50_S2K146_UCA:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S3L17 < `S3L17_censor_lower` | EC50_S3L17 > `S3L17_censor_upper`), EC50_S3L17:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S3L17_UCA < `S3L17_UCA_censor_lower` | EC50_S3L17_UCA > `S3L17_UCA_censor_upper`), EC50_S3L17_UCA:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S309 < `S309_censor_lower` | EC50_S309 > `S309_censor_upper`), EC50_S309:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S309_887 < `S309_887_censor_lower` | EC50_S309_887 > `S309_887_censor_upper`), EC50_S309_887:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_S2X259 < `S2X259_censor_lower` | EC50_S2X259 > `S2X259_censor_upper`), EC50_S2X259:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_SA55 < `SA55_censor_lower` | EC50_SA55 > `SA55_censor_upper`), EC50_SA55:=NA]
dt[!(target %in% config$targets_low_bc) & (EC50_Omi42 < `Omi42_censor_lower` | EC50_Omi42 > `Omi42_censor_upper`), EC50_Omi42:=NA]

#calculate geomean, sd, and # bcs collapsed per genotype
dt[,mean_EC50_S2V29_v37_2:=10^mean(log10(EC50_S2V29_v37_2),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2V29_v37_2:=sd(log10(EC50_S2V29_v37_2),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2V29_v37_2:=sum(!is.na(EC50_S2V29_v37_2)),by=c("library","target")]

dt[,mean_EC50_S2V29:=10^mean(log10(EC50_S2V29),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2V29:=sd(log10(EC50_S2V29),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2V29:=sum(!is.na(EC50_S2V29)),by=c("library","target")]

dt[,mean_EC50_S2V29_UCA:=10^mean(log10(EC50_S2V29_UCA),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2V29_UCA:=sd(log10(EC50_S2V29_UCA),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2V29_UCA:=sum(!is.na(EC50_S2V29_UCA)),by=c("library","target")]

dt[,mean_EC50_S2K146_v860:=10^mean(log10(EC50_S2K146_v860),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2K146_v860:=sd(log10(EC50_S2K146_v860),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2K146_v860:=sum(!is.na(EC50_S2K146_v860)),by=c("library","target")]

dt[,mean_EC50_S2K146:=10^mean(log10(EC50_S2K146),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2K146:=sd(log10(EC50_S2K146),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2K146:=sum(!is.na(EC50_S2K146)),by=c("library","target")]

dt[,mean_EC50_S2K146_UCA:=10^mean(log10(EC50_S2K146_UCA),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2K146_UCA:=sd(log10(EC50_S2K146_UCA),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2K146_UCA:=sum(!is.na(EC50_S2K146_UCA)),by=c("library","target")]

dt[,mean_EC50_S3L17:=10^mean(log10(EC50_S3L17),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S3L17:=sd(log10(EC50_S3L17),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S3L17:=sum(!is.na(EC50_S3L17)),by=c("library","target")]

dt[,mean_EC50_S3L17_UCA:=10^mean(log10(EC50_S3L17_UCA),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S3L17_UCA:=sd(log10(EC50_S3L17_UCA),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S3L17_UCA:=sum(!is.na(EC50_S3L17_UCA)),by=c("library","target")]

dt[,mean_EC50_S309:=10^mean(log10(EC50_S309),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S309:=sd(log10(EC50_S309),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S309:=sum(!is.na(EC50_S309)),by=c("library","target")]

dt[,mean_EC50_S309_887:=10^mean(log10(EC50_S309_887),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S309_887:=sd(log10(EC50_S309_887),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S309_887:=sum(!is.na(EC50_S309_887)),by=c("library","target")]

dt[,mean_EC50_S2X259:=10^mean(log10(EC50_S2X259),na.rm=T),by=c("library","target")]
dt[,sd_EC50_S2X259:=sd(log10(EC50_S2X259),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_S2X259:=sum(!is.na(EC50_S2X259)),by=c("library","target")]

dt[,mean_EC50_SA55:=10^mean(log10(EC50_SA55),na.rm=T),by=c("library","target")]
dt[,sd_EC50_SA55:=sd(log10(EC50_SA55),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_SA55:=sum(!is.na(EC50_SA55)),by=c("library","target")]

dt[,mean_EC50_Omi42:=10^mean(log10(EC50_Omi42),na.rm=T),by=c("library","target")]
dt[,sd_EC50_Omi42:=sd(log10(EC50_Omi42),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_Omi42:=sum(!is.na(EC50_Omi42)),by=c("library","target")]

```

Collapse down to tables reporting just the summary statistics for each genotype.

```{r collapse_to_genotype}
dt_final <- dt[,.(library,target,variant_class,
                  mean_EC50_S2V29_v37_2, sd_EC50_S2V29_v37_2, n_bc_EC50_S2V29_v37_2,
                  mean_EC50_S2V29, sd_EC50_S2V29, n_bc_EC50_S2V29,
                  mean_EC50_S2V29_UCA, sd_EC50_S2V29_UCA, n_bc_EC50_S2V29_UCA,
                  mean_EC50_S2K146_v860, sd_EC50_S2K146_v860, n_bc_EC50_S2K146_v860,
                  mean_EC50_S2K146, sd_EC50_S2K146, n_bc_EC50_S2K146,
                  mean_EC50_S2K146_UCA, sd_EC50_S2K146_UCA, n_bc_EC50_S2K146_UCA,
                  mean_EC50_S3L17, sd_EC50_S3L17, n_bc_EC50_S3L17,
                  mean_EC50_S3L17_UCA, sd_EC50_S3L17_UCA, n_bc_EC50_S3L17_UCA,
                  mean_EC50_S309, sd_EC50_S309, n_bc_EC50_S309,
                  mean_EC50_S309_887, sd_EC50_S309_887, n_bc_EC50_S309_887,
                  mean_EC50_S2X259, sd_EC50_S2X259, n_bc_EC50_S2X259,
                  mean_EC50_SA55, sd_EC50_SA55, n_bc_EC50_SA55,
                  mean_EC50_Omi42, sd_EC50_Omi42, n_bc_EC50_Omi42)]

dt_final <- unique(dt_final); setkey(dt_final, target)
```

Let's look how SEM is distributed. Can see that SEM is generally very, very low. Also that it doesn't really have a relationship with the EC50 metric, which is good.

```{r plot_SEMs, echo=T, fig.width=12, fig.height=12, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(4,4))
#S2V29_v37_2
x <- log10(dt_final[,mean_EC50_S2V29_v37_2]); y <- dt_final[,sd_EC50_S2V29_v37_2/sqrt(n_bc_EC50_S2V29_v37_2)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2V29_v37_2")

#S2V29
x <- log10(dt_final[,mean_EC50_S2V29]); y <- dt_final[,sd_EC50_S2V29/sqrt(n_bc_EC50_S2V29)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2V29")

#S2V29_UCA
x <- log10(dt_final[,mean_EC50_S2V29_UCA]); y <- dt_final[,sd_EC50_S2V29_UCA/sqrt(n_bc_EC50_S2V29_UCA)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2V29_UCA")

#S2K146_v860
x <- log10(dt_final[,mean_EC50_S2K146_v860]); y <- dt_final[,sd_EC50_S2K146_v860/sqrt(n_bc_EC50_S2K146_v860)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2K146_v860")

#S2K146
x <- log10(dt_final[,mean_EC50_S2K146]); y <- dt_final[,sd_EC50_S2K146/sqrt(n_bc_EC50_S2K146)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2K146")

#S2K146_UCA
x <- log10(dt_final[,mean_EC50_S2K146_UCA]); y <- dt_final[,sd_EC50_S2K146_UCA/sqrt(n_bc_EC50_S2K146_UCA)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2K146_UCA")

#S3L17
x <- log10(dt_final[,mean_EC50_S3L17]); y <- dt_final[,sd_EC50_S3L17/sqrt(n_bc_EC50_S3L17)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S3L17")

#S3L17_UCA
x <- log10(dt_final[,mean_EC50_S3L17_UCA]); y <- dt_final[,sd_EC50_S3L17_UCA/sqrt(n_bc_EC50_S3L17_UCA)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S3L17_UCA")

#S309
x <- log10(dt_final[,mean_EC50_S309]); y <- dt_final[,sd_EC50_S309/sqrt(n_bc_EC50_S309)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S309")

#S309_887
x <- log10(dt_final[,mean_EC50_S309_887]); y <- dt_final[,sd_EC50_S309_887/sqrt(n_bc_EC50_S309_887)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S309_887")

#S2X259
x <- log10(dt_final[,mean_EC50_S2X259]); y <- dt_final[,sd_EC50_S2X259/sqrt(n_bc_EC50_S2X259)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="S2X259")

#SA55
x <- log10(dt_final[,mean_EC50_SA55]); y <- dt_final[,sd_EC50_SA55/sqrt(n_bc_EC50_SA55)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="SA55")

#Omi42
x <- log10(dt_final[,mean_EC50_Omi42]); y <- dt_final[,sd_EC50_Omi42/sqrt(n_bc_EC50_Omi42)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="Omi42")


invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/SEM-v-EC50.pdf",sep=""),useDingbats=F))
```


Add in the previously-measured expression values.

```{r add_expr, fig.width=6, fig.height=12, fig.align="center", dpi=300,dev="png"}
dt_final[,expr:=as.numeric(NA)]
for(i in 1:nrow(dt_final)){
  bg <- as.character(dt_final[i,target])
  if(bg %in% dt_expr$target){
    dt_final[i,expr := dt_expr[target==bg,expression]]
  }
}

#compute a delta_expr relative to the median
dt_final[,delta_expr := expr - median(dt_final$expr,na.rm=T)]


```


Filter out the two backgrounds that were completely non-expressing. Most barcodes were purged before the affinity measurements for these backgrounds, so the affinities are determined from few barcodes and are just generally unreliable because these are poorly folded/expressing variants. (E.g. could see very high standard deviations)

```{r remove_nonexpressing}
dt_final[target %in% c("HKU3-8","AncSARS1a_alt"),c("mean_EC50_S2V29_v37_2","sd_EC50_S2V29_v37_2","n_bc_EC50_S2V29_v37_2",
                                                   "mean_EC50_S2V29","sd_EC50_S2V29","n_bc_EC50_S2V29",
                                                   "mean_EC50_S2V29_UCA","sd_EC50_S2V29_UCA","n_bc_EC50_S2V29_UCA",
                                                   "mean_EC50_S2K146_v860","sd_EC50_S2K146_v860","n_bc_EC50_S2K146_v860",
                                                   "mean_EC50_S2K146","sd_EC50_S2K146","n_bc_EC50_S2K146",
                                                   "mean_EC50_S2K146_UCA","sd_EC50_S2K146_UCA","n_bc_EC50_S2K146_UCA",
                                                   "mean_EC50_S3L17","sd_EC50_S3L17","n_bc_EC50_S3L17",
                                                   "mean_EC50_S3L17_UCA","sd_EC50_S3L17_UCA","n_bc_EC50_S3L17_UCA",
                                                   "mean_EC50_S309","sd_EC50_S309","n_bc_EC50_S309",
                                                   "mean_EC50_S309_887","sd_EC50_S309_887","n_bc_EC50_S309_887",
                                                   "mean_EC50_S2X259","sd_EC50_S2X259","n_bc_EC50_S2X259",
                                                   "mean_EC50_SA55","sd_EC50_SA55","n_bc_EC50_SA55",
                                                   "mean_EC50_Omi42","sd_EC50_Omi42","n_bc_EC50_Omi42"):=NA]
```

Coverage stats on n_barcodes for different measurements in the final pooled measurements.

```{r n_barcode_plots, echo=T, fig.width=16, fig.height=12, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(4,4))

hist(dt_final$n_bc_EC50_S2V29_v37_2,col="gray50",main=paste("S2V29_v37_2,\nmedian ",median(dt_final$n_bc_EC50_S2V29_v37_2,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2V29,col="gray50",main=paste("S2V29,\nmedian ",median(dt_final$n_bc_EC50_S2V29,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2V29_UCA,col="gray50",main=paste("S2V29_UCA,\nmedian ",median(dt_final$n_bc_EC50_S2V29_UCA,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2K146_v860,col="gray50",main=paste("S2K146_v860,\nmedian ",median(dt_final$n_bc_EC50_S2K146_v860,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2K146,col="gray50",main=paste("S2K146,\nmedian ",median(dt_final$n_bc_EC50_S2K146,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2K146_UCA,col="gray50",main=paste("S2K146_UCA,\nmedian ",median(dt_final$n_bc_EC50_S2K146_UCA,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S3L17,col="gray50",main=paste("S3L17,\nmedian ",median(dt_final$n_bc_EC50_S3L17,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S3L17_UCA,col="gray50",main=paste("S3L17_UCA,\nmedian ",median(dt_final$n_bc_EC50_S3L17_UCA,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S309,col="gray50",main=paste("S309,\nmedian ",median(dt_final$n_bc_EC50_S309,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S309_887,col="gray50",main=paste("S309_887,\nmedian ",median(dt_final$n_bc_EC50_S309_887,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_S2X259,col="gray50",main=paste("S2X259,\nmedian ",median(dt_final$n_bc_EC50_S2X259,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_SA55,col="gray50",main=paste("SA55,\nmedian ",median(dt_final$n_bc_EC50_SA55,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_Omi42,col="gray50",main=paste("Omi42,\nmedian ",median(dt_final$n_bc_EC50_Omi42,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_hist_n_barcodes.pdf",sep="")))
```
Order factor variables for plotting

```{r order_plotting_factors}
#order target by order given in config
dt_final$target <- factor(dt_final$target,levels=config$targets_ordered)

#rename some columns for convenience
setnames(dt_final,"mean_EC50_S2V29_v37_2","EC50_S2V29_v37_2")
setnames(dt_final,"mean_EC50_S2V29","EC50_S2V29")
setnames(dt_final,"mean_EC50_S2V29_UCA","EC50_S2V29_UCA")
setnames(dt_final,"mean_EC50_S2K146_v860","EC50_S2K146_v860")
setnames(dt_final,"mean_EC50_S2K146","EC50_S2K146")
setnames(dt_final,"mean_EC50_S2K146_UCA","EC50_S2K146_UCA")
setnames(dt_final,"mean_EC50_S3L17","EC50_S3L17")
setnames(dt_final,"mean_EC50_S3L17_UCA","EC50_S3L17_UCA")
setnames(dt_final,"mean_EC50_S309","EC50_S309")
setnames(dt_final,"mean_EC50_S309_887","EC50_S309_887")
setnames(dt_final,"mean_EC50_S2X259","EC50_S2X259")
setnames(dt_final,"mean_EC50_SA55","EC50_SA55")
setnames(dt_final,"mean_EC50_Omi42","EC50_Omi42")

```

## Heatmaps! 

Output heatmaps illustrating all wildtype variants with separate columns for each mAb.

```{r heatmap_wildtypes_all_EC50, fig.width=12,fig.height=5.5,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
temp1 <- data.table::melt(dt_final[,.(target, EC50_S309, EC50_S309_887, EC50_S3L17_UCA, EC50_S3L17, EC50_S2K146_UCA,EC50_S2K146,EC50_S2K146_v860,EC50_S2V29_UCA,EC50_S2V29 ,EC50_S2V29_v37_2, EC50_SA55, EC50_Omi42, EC50_S309)],
                          id.vars=c("target"),
                          measure.vars=c("EC50_S309", "EC50_S309_887", "EC50_S3L17_UCA", "EC50_S3L17", "EC50_S2K146_UCA","EC50_S2K146","EC50_S2K146_v860","EC50_S2V29_UCA","EC50_S2V29", "EC50_S2V29_v37_2", "EC50_SA55", "EC50_Omi42", "EC50_S309"),
                          variable.name="mAb",value.name="EC50")

temp1$log10_EC50 <- log10(temp1$EC50)

p1 <- ggplot(temp1,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-2,5),values=c(0,7/7,7/7),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

grid.arrange(p1,nrow=1)

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_all_wildtypes.pdf",sep="")))

```
Showing just extant sarbs.

```{r heatmap_wildtypes_extants_EC50, fig.width=12,fig.height=5.5,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
extant <- c(config$EurAf_extant,config$RsYN04_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp2 <- temp1[target %in% extant,];temp2$target <- factor(temp2$target,levels=extant)

p1 <- ggplot(temp2,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-2,5),values=c(0,7/7,7/7),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

p1

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_extant_wildtypes.pdf",sep="")))

```

And for ancestors.

```{r heatmap_wildtypes_MAP-ancestors_EC50, fig.width=8,fig.height=5.5,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
ancestors <- c(config$ancestors_MAP)

temp3 <- temp1[target %in% ancestors,];temp3$target <- factor(temp3$target,levels=ancestors)

p1 <- ggplot(temp3,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-2,5),values=c(0,7/7,7/7),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

p1

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_MAP-ancestors.pdf",sep="")))

```


Save output file.

```{r outputs}
dt_final %>%
  mutate_if(is.numeric, round, digits=5) %>%
  write.csv(file=config$final_variant_scores_lib61_file, row.names=F,quote=F)

```





