---
title: "Collapse barcodes to final per-RBD/mutant phenotype scores for the wildtype sarbecoviruses pool"
author: "Tyler Starr"
date: "12/28/2022"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---
This notebook reads in the per-barcode sera binding values and previously measured expression for sarbecovirus homologs pool. It synthesizes these two sets of results and calculates the final 'mean' phenotypes for each variant, and generates some coverage and QC analyses.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages],
                   lib=c(paste("/uufs/chpc.utah.edu/common/home/",Sys.getenv("USER"),"/RLibs/",Sys.getenv("R_VERSION"),sep="")),
                   repos=c("http://cran.us.r-project.org"))
}
#load packages
invisible(lapply(packages, library, character.only=T))

knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$final_variant_scores_dir)){
  dir.create(file.path(config$final_variant_scores_dir))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

## Setup

Read in tables of per-barcode AUC

```{r input_data}
dt <- data.table(read.csv(config$mAb_EC50_file),stringsAsFactors=F)[library=="lib61_SARSr-wts",]

#check all targets are in the targets_ordered config list
unique(dt$target) %in% config$targets_ordered

#assign target as a factor in my desired overall plotting order
dt[,target := factor(dt$target,levels=config$targets_ordered)]

#read in previously measured expression measurements for these variants? Note, don't have this metric for the pool7 additions to this v2 pool
dt_expr <- data.table(read.csv(config$SARSr_lib61_mut_bind_expr),stringsAsFactors=F)
dt_expr[target=="SARS-CoV-2",target:="SARS-CoV-2_WH1"]
```

## Calculate per-variant mean scores

Unfiltered, look at distribution of AUC scores

```{r unfiltered_AUCs, echo=T, fig.width=10, fig.height=18, fig.align="center", dpi=300,dev="png"}
p1 <- ggplot(dt[!is.na(EC50_268C61),],aes(x=target,y=log10(EC50_268C61)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C61 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p2 <- ggplot(dt[!is.na(EC50_268C183),],aes(x=target,y=log10(EC50_268C183)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C183 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p3 <- ggplot(dt[!is.na(EC50_268C185),],aes(x=target,y=log10(EC50_268C185)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C185 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p4 <- ggplot(dt[!is.na(EC50_268C3),],aes(x=target,y=log10(EC50_268C3)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C3 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

grid.arrange(p1,p2,p3,p4,ncol=1)
```

Let's add a variable that flags the top and bottom 5% of expression scores for each variant, and see how violin plots look when censoring these top and bottom 10% of expressed barcodes

```{r censor_2.5_AUCs, echo=T, fig.width=10, fig.height=16, fig.align="center", dpi=300,dev="png"}
dt[,`268C61_censor_lower`:=quantile(EC50_268C61,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`268C61_censor_upper`:=quantile(EC50_268C61,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`268C183_censor_lower`:=quantile(EC50_268C183,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`268C183_censor_upper`:=quantile(EC50_268C183,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`268C185_censor_lower`:=quantile(EC50_268C185,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`268C185_censor_upper`:=quantile(EC50_268C185,0.95,na.rm=T,type=7),by=c("library","target")]

dt[,`268C3_censor_lower`:=quantile(EC50_268C3,0.05,na.rm=T,type=7),by=c("library","target")]
dt[,`268C3_censor_upper`:=quantile(EC50_268C3,0.95,na.rm=T,type=7),by=c("library","target")]

p1 <- ggplot(dt[!is.na(EC50_268C61),],aes(x=target,y=log10(EC50_268C61)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C61 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p2 <- ggplot(dt[!is.na(EC50_268C183),],aes(x=target,y=log10(EC50_268C183)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C183 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p3 <- ggplot(dt[!is.na(EC50_268C185),],aes(x=target,y=log10(EC50_268C185)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C185 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

p4 <- ggplot(dt[!is.na(EC50_268C3),],aes(x=target,y=log10(EC50_268C3)))+
  geom_violin(scale="width")+stat_summary(fun=median,geom="point",size=1)+
  ggtitle("268C3 mAb binding")+xlab("homolog")+theme(axis.text.x=element_text(angle=-90,hjust=0))

grid.arrange(p1,p2,p3,p4,ncol=1)
invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_vioplots_AUC-cens.pdf",sep="")))
```

Calculate the mean (geometric mean) per variant, the standard deviation, and the number of (post-filter) barcodes on which a variant score was determined

````{r calculate_mean_scores}
#apply the censors to NA out the phenotypes outside the range
dt[EC50_268C61 < `268C61_censor_lower` | EC50_268C61 > `268C61_censor_upper`, EC50_268C61:=NA]
dt[EC50_268C183 < `268C183_censor_lower` | EC50_268C183 > `268C183_censor_upper`, EC50_268C183:=NA]
dt[EC50_268C185 < `268C185_censor_lower` | EC50_268C185 > `268C185_censor_upper`, EC50_268C185:=NA]
dt[EC50_268C3 < `268C3_censor_lower` | EC50_268C3 > `268C3_censor_upper`, EC50_268C3:=NA]

dt[,mean_EC50_268C61:=10^mean(log10(EC50_268C61),na.rm=T),by=c("library","target")]
dt[,sd_EC50_268C61:=sd(log10(EC50_268C61),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_268C61:=sum(!is.na(EC50_268C61)),by=c("library","target")]

dt[,mean_EC50_268C183:=10^mean(log10(EC50_268C183),na.rm=T),by=c("library","target")]
dt[,sd_EC50_268C183:=sd(log10(EC50_268C183),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_268C183:=sum(!is.na(EC50_268C183)),by=c("library","target")]

dt[,mean_EC50_268C185:=10^mean(log10(EC50_268C185),na.rm=T),by=c("library","target")]
dt[,sd_EC50_268C185:=sd(log10(EC50_268C185),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_268C185:=sum(!is.na(EC50_268C185)),by=c("library","target")]

dt[,mean_EC50_268C3:=10^mean(log10(EC50_268C3),na.rm=T),by=c("library","target")]
dt[,sd_EC50_268C3:=sd(log10(EC50_268C3),na.rm=T),by=c("library","target")]
dt[,n_bc_EC50_268C3:=sum(!is.na(EC50_268C3)),by=c("library","target")]

```

Collapse down to tables reporting just the summary statistics for each genotype.

```{r collapse_to_genotype}
dt_final <- dt[,.(library,target,variant_class,
                  mean_EC50_268C61, sd_EC50_268C61, n_bc_EC50_268C61,
                  mean_EC50_268C183, sd_EC50_268C183, n_bc_EC50_268C183,
                  mean_EC50_268C185, sd_EC50_268C185, n_bc_EC50_268C185,
                  mean_EC50_268C3, sd_EC50_268C3, n_bc_EC50_268C3)]

dt_final <- unique(dt_final); setkey(dt_final, target)
```

Let's look how SEM is distributed. Can see that SEM is generally very, very low. Also that it doesn't really have a relationship with the EC50 metric, which is good.

```{r plot_SEMs, echo=T, fig.width=6, fig.height=6, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(2,2))
#268C61
x <- log10(dt_final[,mean_EC50_268C61]); y <- dt_final[,sd_EC50_268C61/sqrt(n_bc_EC50_268C61)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="268C61")

#268C183
x <- log10(dt_final[,mean_EC50_268C183]); y <- dt_final[,sd_EC50_268C183/sqrt(n_bc_EC50_268C183)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="268C183")

#268C185
x <- log10(dt_final[,mean_EC50_268C185]); y <- dt_final[,sd_EC50_268C185/sqrt(n_bc_EC50_268C185)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="268C185")

#268C3
x <- log10(dt_final[,mean_EC50_268C3]); y <- dt_final[,sd_EC50_268C3/sqrt(n_bc_EC50_268C3)]; plot(x,y,pch=16,col="#00000090",xlab="EC50",ylab="SEM",main="268C3")

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/SEM-v-EC50.pdf",sep=""),useDingbats=F))
```


Add in the previously-measured expression values.

```{r add_expr, fig.width=6, fig.height=12, fig.align="center", dpi=300,dev="png"}
dt_final[,expr:=as.numeric(NA)]
for(i in 1:nrow(dt_final)){
  bg <- as.character(dt_final[i,target])
  if(bg %in% dt_expr$target){
    dt_final[i,expr := dt_expr[target==bg,expression]]
  }
}

#compute a delta_expr relative to the median
dt_final[,delta_expr := expr - median(dt_final$expr,na.rm=T)]


```


Filter out the two backgrounds that were completely non-expressing. Most barcodes were purged before the affinity measurements for these backgrounds, so the affinities are determined from few barcodes and are just generally unreliable because these are poorly folded/expressing variants. (E.g. could see very high standard deviations)

```{r remove_nonexpressing}
dt_final[target %in% c("HKU3-8","AncSARS1a_alt"),c("mean_EC50_268C61","sd_EC50_268C61","n_bc_EC50_268C61",
                                                              "mean_EC50_268C183","sd_EC50_268C183","n_bc_EC50_268C183",
                                                              "mean_EC50_268C185","sd_EC50_268C185","n_bc_EC50_268C185",
                                                              "mean_EC50_268C3","sd_EC50_268C3","n_bc_EC50_268C3"):=NA]
```

Coverage stats on n_barcodes for different measurements in the final pooled measurements.

```{r n_barcode_plots, echo=T, fig.width=8, fig.height=6, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(2,2))

hist(dt_final$n_bc_EC50_268C61,col="gray50",main=paste("268C61,\nmedian ",median(dt_final$n_bc_EC50_268C61,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_268C183,col="gray50",main=paste("268C183,\nmedian ",median(dt_final$n_bc_EC50_268C183,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_268C185,col="gray50",main=paste("268C185,\nmedian ",median(dt_final$n_bc_EC50_268C185,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

hist(dt_final$n_bc_EC50_268C3,col="gray50",main=paste("268C3,\nmedian ",median(dt_final$n_bc_EC50_268C3,na.rm=T),sep=""),xlab="number barcodes",ylab="number genotypes",breaks=20)

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_hist_n_barcodes.pdf",sep="")))
```
Order factor variables for plotting

```{r order_plotting_factors}
#order target by order given in config
dt_final$target <- factor(dt_final$target,levels=config$targets_ordered)

#rename some columns for convenience
setnames(dt_final,"mean_EC50_268C61","EC50_268C61")
setnames(dt_final,"mean_EC50_268C183","EC50_268C183")
setnames(dt_final,"mean_EC50_268C185","EC50_268C185")
setnames(dt_final,"mean_EC50_268C3","EC50_268C3")
```

## Heatmaps! 

Output heatmaps illustrating all wildtype variants with separate columns for each serum. Do with both the raw AUC and the expression-normalized metric.

```{r heatmap_wildtypes_all_EC50, fig.width=12,fig.height=4,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
temp1 <- data.table::melt(dt_final[,.(target,EC50_268C61,EC50_268C183,EC50_268C185,EC50_268C3)],
                          id.vars=c("target"),
                          measure.vars=c("EC50_268C61","EC50_268C183","EC50_268C185","EC50_268C3"),
                          variable.name="mAb",value.name="EC50")

temp1$log10_EC50 <- log10(temp1$EC50)

p1 <- ggplot(temp1,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-1,5),values=c(0,6/6,6/6),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

grid.arrange(p1,nrow=1)

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_all_wildtypes.pdf",sep="")))

```
Showing just extant sarbs.

```{r heatmap_wildtypes_extants_EC50, fig.width=12,fig.height=4,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
extant <- c(config$EurAf_extant,config$RsYN04_extant,config$SARS2_extant,config$SARS1_extant,config$Clade2_extant)

temp2 <- temp1[target %in% extant,];temp2$target <- factor(temp2$target,levels=extant)

p1 <- ggplot(temp2,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-1,5),values=c(0,6/6,6/6),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

p1

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_extant_wildtypes.pdf",sep="")))

```

And for ancestors.

```{r heatmap_wildtypes_MAP-ancestors_EC50, fig.width=8,fig.height=4,fig.align="center", dpi=500,dev="png",echo=T}
#make temp long-form data frame
ancestors <- c(config$ancestors_MAP)

temp3 <- temp1[target %in% ancestors,];temp3$target <- factor(temp3$target,levels=ancestors)

p1 <- ggplot(temp3,aes(target,mAb))+geom_tile(aes(fill=log10_EC50),color="black",lwd=0.1)+
  scale_fill_gradientn(colours=c("#003366","#FFFFFF","#FFFFFF"),limits=c(-1,5),values=c(0,6/6,6/6),na.value="gray40")+
  #scale_fill_gradientn(colours=c("#FFFFFF","#003366"),limits=c(5,12),values=c(0,1),na.value="yellow")+
  #scale_x_continuous(expand=c(0,0),breaks=c(331,seq(335,430,by=5)))+
  labs(x="RBD homolog",y="")+theme_classic(base_size=9)+
  coord_equal()+theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold"))

p1

invisible(dev.print(pdf, paste(config$final_variant_scores_dir,"/lib61_heatmap_EC50_MAP-ancestors.pdf",sep="")))

```


Save output file.

```{r outputs}
dt_final %>%
  mutate_if(is.numeric, round, digits=5) %>%
  write.csv(file=config$final_variant_scores_lib61_file, row.names=F,quote=F)

```





